### PR TITLE
Group tooltips for line chart

### PR DESCRIPTION
**Pull request by Steve-Mcl**
## Description

Groups tooltips on `line` type charts using the [interactions](https://www.chartjs.org/docs/latest/samples/tooltip/interactions.html) options 

NOTES:
* I wrapped the call to `this.chart.update` into a `chartUpdate` method so that we can debounce the multiple calls generated by property updates (the props are `watch`ed individually) 
  * I then noted we dont actually support dynamic prop updates 🤦‍♂️ BUT I left it in as dynamic updates are listed as a TODO and this will avoid unnecessary chart update calls.
* I added some debugging where I add the `chart` object to the `window` object so that during debugging, I can see the actual config and data applied to the chart. This was super useful. So instead of deleting it, I let it in commented with a note for future devs to take advantage of.  fee free to remove it.

## Related Issue(s)

#1254

## Checklist

 - [x] I have read the [contribution guidelines](https://github.com/FlowFuse/flowfuse/blob/main/CONTRIBUTING.md)
 - [ ] Suitable unit/system level tests have been added and they pass 
 - [ ] Documentation has been updated
    - [ ] Upgrade instructions
    - [ ] Configuration details
    - [ ] Concepts
 - [ ] Changes `flowforge.yml`?
    - [ ] Issue/PR raised on `FlowFuse/helm` to update ConfigMap Template
    - [ ] Issue/PR raised on `FlowFuse/CloudProject` to update values for Staging/Production

## Labels

 - [ ] Includes a DB migration? -> add the `area:migration` label

### CODE REVIEW

**Comment by Steve-Mcl** [comment]
NOTE: I will add tests for this once #1331 is merged.

File: cypress/tests/widgets/chart-options.spec.js [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+        cy.deployFixture('dashboard-chart-options', overrides)
+        cy.visit('/dashboard/page1')
+        cy.reload()
+
+        cy.get('#nrdb-ui-widget-chart-1 > div > canvas').should('exist')
+        cy.clickAndWait(cy.get('button').contains('random'), 10)
+
+        // eslint-disable-next-line promise/catch-or-return, promise/always-return
+        cy.window().then(win => {
+            should(win.uiCharts).is.not.empty()
```
Comment (joepavitt) [inline cypress/tests/widgets/chart-options.spec.js:57]: If `win.uiCharts` is already available, then why also the need for `window.uiChart` which is commented out?

File: ui/src/widgets/ui-chart/UIChart.vue [hunk 2]
```diff
@@ -191,10 +194,52 @@ export default {
...
+        updateInteraction () {
+            switch (this.chart.config.type) {
+            case 'line':
+                this.chart.options.interaction.axis = 'x'
+                this.chart.options.interaction.mode = 'index'
+                break
+            default:
+                delete this.chart.options.interaction.axis
+                this.chart.options.interaction.mode = 'nearest' // default
+                break
```
Comment (joepavitt) [inline ui/src/widgets/ui-chart/UIChart.vue:218]: I mentioned [here](https://github.com/FlowFuse/node-red-dashboard/issues/1254#issuecomment-2371774649) that `mode` should be `x` for the line charts, as if you have different lines with a different number of data points, then the tooltips get mis-aligned.

```suggestion
            case 'line':
                this.chart.options.interaction.mode = 'x'
                break
            case 'scatter':
                this.chart.options.interaction.axis = 'x'
                this.chart.options.interaction.mode = 'nearest'
                break
            default:
                delete this.chart.options.interaction.axis
                this.chart.options.interaction.mode = 'index' // default
                break
```

Having played around too, the bar and donut charts benefit from `index` type, with the scatter reverting to the `nearest` tooltip type.

File: cypress/tests/widgets/chart-options.spec.js (hunk 1 above)
Comment (Steve-Mcl) [reply cypress/tests/widgets/chart-options.spec.js:57]: Not sure I understand the question Joe.

Where is window.uiChart commented out?

For reference: `win.uiCharts` is added explicitly in `UIChart.vue` when `Cypress` is detected (only). 
This is pretty much the only way of getting to the end for E2E tests. It was debated and researched with Serban at the time of impleneting the first round of e2e tests for charts.

File: ui/src/widgets/ui-chart/UIChart.vue (hunk 2 above)
Comment (Steve-Mcl) [reply ui/src/widgets/ui-chart/UIChart.vue:218]: > I mentioned https://github.com/FlowFuse/node-red-dashboard/issues/1254#issuecomment-2371774649 that mode should be x for the line charts


I misread the API docs https://www.chartjs.org/docs/latest/configuration/interactions.html#modes (didnt scroll down enough to see there was a mode 'x' and thought what you entered was a typo. My bad!)

File: cypress/tests/widgets/chart-options.spec.js (hunk 1 above)
Comment (Steve-Mcl) [reply cypress/tests/widgets/chart-options.spec.js:57]: > window.uiChart which is commented out?

ah, you mean in the other file - in UIChart.vue:
https://github.com/FlowFuse/node-red-dashboard/pull/1336/files/6040f1ca6c988150cb901e0d1292552e7e925614

Those commented out lines were super helpful during development of tests (so I could see what was set in the chart  object).

the line you have commented on here is related directly to existing code here
https://github.com/FlowFuse/node-red-dashboard/pull/1336/files/6040f1ca6c988150cb901e0d1292552e7e925614#diff-0ab94e22f1954c12fd38be2754bb91b4c72e3ad50b23742c57a12ff9c3e9e194R60-R65

As mentioned above, `uiChart` is conditionally added to the window object ONLY when cypress is detected. This highlighted code (in the test runner) picks up the chart object to get as close to the end as possible.

**Comment by Steve-Mcl** [comment]
Merged your correction to the `mode` and addition for scatter. Ta.
I will need to update tests to pick up additional grouping chart type for the above.

Addressed you comment about `window.uiChart` in comment here: https://github.com/FlowFuse/node-red-dashboard/pull/1336#discussion_r1783199460

**Comment by Steve-Mcl** [comment]
Tests updated/added to cover changes for scatter and bar.

All passing bar one of the control spec tests (usual suspect) (really would love to get that reliable!)

**Comment by joepavitt** [comment]
> (really would love to get that reliable!)

@Steve-Mcl - @cstns and I have collectively spent days trying - we've given up.

**Comment by joepavitt** [comment]
Works great - thanks @Steve-Mcl

**Review by joepavitt — APPROVED** [review]

**Pull request merged**